validations/src/validations.rs

```diff
@@ -215,55 +220,74 @@ fn update_liars(liars: &mut Vec<bool>, item: RadonTypes, condition: bool) -> Opt
     }
 }
 
+/// An `Either`-like structure that covers the two possible return types of the
+/// `evaluate_tally_precondition_clause` method.
+#[derive(Debug)]
+pub enum TallyPreconditionClauseResult {
+    MajorityOfValues(Vec<RadonTypes>, Vec<bool>),
+    MajorityOfErrors(RadonTypes),
+}
+
+/// Run a precondition clause on an array of `RadonTypes` so as to check if the mode is a value or
+/// an error, which has clear consequences in regards to consensus, rewards and punishments.
 pub fn evaluate_tally_precondition_clause(
     reveals: Vec<RadonReport<RadonTypes>>,
     non_error_min: f64,
-) -> Result<(Vec<RadonTypes>, Vec<bool>), RadError> {
+) -> Result<TallyPreconditionClauseResult, RadError> {
+    // Short-circuit if there were no reveals
     if reveals.is_empty() {
         return Err(RadError::NoReveals);
     }
 
     let reveals_len = reveals.len() as f64;
-
     let mut counter = Counter::new(RadonTypes::num_types());
     for reveal in &reveals {
-        match &reveal.result {
-            Err(_e) => {}
-            Ok(rad_types) => counter.increment(rad_types.discriminant()),
-        }
+        counter.increment(reveal.result.discriminant());
+    }
+
+    let error_ratio = counter.values
+        [RadonTypes::RadonError(RadonError::from(RadonErrors::default())).discriminant()]
+        as f64
+        / reveals_len;
+
+    if counter.max_pos.is_none() {
+        return Err(RadError::ModeTie {
+            values: RadonArray::from(
+                reveals
+                    .into_iter()
+                    .map(RadonReport::into_inner)
+                    .collect::<Vec<RadonTypes>>(),
+            ),
+        });
     }
 
-    let non_error_ratio = counter.max_val as f64 / reveals_len;
-
-    if non_error_ratio > non_error_min {
+    if error_ratio <= non_error_min {
         let mut liars = vec![];
         let results = reveals
             .into_iter()
-            .filter_map(|reveal| match reveal.into_inner() {
-                Err(_e) => {
-                    liars.push(true);
-                    None
-                }
-                Ok(rad_types) => {
-                    let condition = counter.max_pos == rad_types.discriminant();
-                    update_liars(&mut liars, rad_types, condition)
-                }
+            .filter_map(|reveal| {
+                let radon_types = reveal.into_inner();
+                let condition = counter.max_pos.unwrap() == radon_types.discriminant();
```

Comment:
What happens if `counter.max_pos` is the discriminant of `RadonError`? For example:

10 reveals: 4 RadonError, 1 RadonInteger, 1 RadonFloat, 1 RadonString, 1 RadonBoolean, 1 RadonMap, 1 RadonArray.

error_ratio = 4/10 which is smaller than 0.51

But the most common type is `RadonError`.

Add test with this case.

Reply:
Here's a simpler example of what I mean

```rust
    #[test]
    fn test_tally_precondition_clause_error3() {
        let rad_int = RadonTypes::Integer(RadonInteger::from(1));
        let rad_float = RadonTypes::Float(RadonFloat::from(1));
        let rad_err = RadonError::from(RadonErrors::HTTPError);

        let rad_rep_int = RadonReport::from_result(Ok(rad_int), &ReportContext::default()).unwrap();
        let rad_rep_float =
            RadonReport::from_result(Ok(rad_float), &ReportContext::default()).unwrap();
        let rad_rep_err = RadonReport::from_result(
            Ok(RadonTypes::RadonError(rad_err)),
            &ReportContext::default(),
        )
        .unwrap();

        let v = vec![
            rad_rep_err.clone(),
            rad_rep_err.clone(),
            rad_rep_err,
            rad_rep_float.clone(),
            rad_rep_int.clone(),
            rad_rep_float,
            rad_rep_int,
        ];

        let out = evaluate_tally_precondition_clause(v.clone(), 0.51).unwrap_err();

        assert_eq!(
            out,
            RadError::ModeTie {
                values: RadonArray::from(
                    v.into_iter()
                        .map(RadonReport::into_inner)
                        .collect::<Vec<RadonTypes>>()
                )
            }
        );
    }
```

Reply:
Done!